app/services/hyrax/edit_permissions_service.rb

```diff
@@ -0,0 +1,178 @@
+# frozen_string_literal: true
+module Hyrax
+  # Encapsulates the logic to determine which object permissions may be edited by a given user
+  #  - user is permitted to update any work permissions coming ONLY from collections they manage
+  #  - user is not permitted to update a work permission if it comes from a collection they do not manage, even if also from a managed collection
+  #  - user is permitted to update only non-manager permissions from any Collections
+  #  - user is permitted to update any non-collection permissions
+  class EditPermissionsService
+    attr_reader :depositor, :unauthorized_collection_managers
+
+    # @param [Object] GenericWorkForm (if called for object) or GenericWork (if called for file set)
```

Comment:
is the interface `@param object [#depositor, #admin_set_id, #member_of_collection_ids]`?